### DISCUSSION TITLE
`readonly struct` Should not contain reference type

### DISCUSSION THREAD

**Discussion by MohamadAlallan-Pi** [dotnet/roslyn#75249] [Ideas] [answered]
Reference types in `readonly struct` breaks the design of immutable struct.

This code works pretty good as the `readonly struct` has `value types` only
```csharp
static void Main(string[] args)
    {
        var b = new Bar { Value = "bar before foo" };
        var f = new Foo(b);
        Console.WriteLine(f.Value);
        b.Value = "bar after foo";

        f.Bar.Value = "bar from f.Bar";

        Console.WriteLine(f.Value);
    }
    readonly struct Foo
    {
        public string Value => Bar.Value;
        public Bar Bar { get; }
        public Foo(Bar b) => Bar = b;
    }
    struct Bar
    {
        public string Value { get; set; }
    }
```

but once you change `Bar` to a **Reference Type: class Bar**, things will break, you will have two issues
* The compiler error at line `f.Bar.Value = "bar from f.Bar";` will disappear allowing the code to compile
* The f.Value is not readonly anymore as it can be manipulated by any reference to Bar (unlike when Bar was copied by value so other reference have no effect)

Having that, the compiler should prevent reference types from being used in `readonly struct`

**Comment by CyrusNajmabadi** [comment, marked as answer]
Read-only structs are not immutable, they are read-only.  This has been the design of read-only in the language since 1.0.  Changing this would break the world. 

If you want to have this restriction in your own code, you can write an analyzer for that purpose.

**Reply by miloush** [reply]
I will point out that the [C# reference for structs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#readonly-struct) opens with

> You use the readonly modifier to declare that a structure type is immutable.
